Add support for Ubuntu 22.04

stefaniapedrazzi opened the pull request:
Add support of the latest Ubuntu 22.04.

* [x] Remove specific code for automatically installing optional dependencies and creating packages on Ubuntu 18.04.
* [x] Remove code to compile Python3.6 API (not available by default on Ubuntu 22.04)
* [x] On Ubuntu 22.04, download OpenSSL dependencies from Cyberbotics server
* [x] Create and upload new sumo package for Ubuntu 20.04
* [x] Update instructions to create packages on Ubuntu 22.04
* [x] Update Ubuntu 20.04 package creation to make them work on Ubuntu 22.04 too

Notes:
* ROS is currently not officially supported on Ubuntu 22.04 and it is not known if it will be in the future given that *ROS noetic* is the last release.

Errors fixed:
* ``Warning: Incompatible version of OpenSSL (built with OpenSSL 1.x, runtime version is >= 3.x)``
   -> Default SSL library is version 3 but we Qt6 is still built with version 1.
   We need to re-add the download of default Ubuntu 20.04 system OpenSSL libraries.
* Sumo interface: changed default version of `libgdal` and `libproj`

Tests:
* run Ubuntu 20.04 packages:
    * [x] Debian package
    * [x] tarball -> works after including openssl libraries and additional sumo libraries in the packages.
* local package creation
    * [x] tarball

stefaniapedrazzi commented:
It remains just one issue with the Ubuntu 22.04 tests on the CI where the robot window test fails.
Note that the Ubuntu 22.04 environment is still in beta and for example it seems that Firefox is not installed yet.

I tried to force using chromium instead but it didn't solve the issue.

I probably think it is better to wait until the Ubuntu 22.04 environment for the GitHub action is fully released before further inspecting the issue. For the moment we could simply skip the robot window test and run all the others.

stefaniapedrazzi commented:
> After cleansing and recompiling the develop branch I noticed that python 3.10 throws a deprecation warning the first time around:
> 
> ```
> controller310.cpp: In function ‘PyObject* PyInit__controller()’:
> controller310.cpp:980:65: warning: ‘void PyEval_InitThreads()’ is deprecated [-Wdeprecated-declarations]
>   980 | #     define SWIG_PYTHON_INITIALIZE_THREADS  PyEval_InitThreads()
>       |                                                                 ^
> controller310.cpp:27431:3: note: in expansion of macro ‘SWIG_PYTHON_INITIALIZE_THREADS’
> 27431 |   SWIG_PYTHON_INITIALIZE_THREADS;
>       |   ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
> In file included from /usr/include/python3.10/Python.h:130,
>                  from controller310.cpp:182:
> /usr/include/python3.10/ceval.h:122:37: note: declared here
>   122 | Py_DEPRECATED(3.9) PyAPI_FUNC(void) PyEval_InitThreads(void);
>       |                                     ^~~~~~~~~~~~~~~~~~
> controller310.cpp:980:65: warning: ‘void PyEval_InitThreads()’ is deprecated [-Wdeprecated-declarations]
>   980 | #     define SWIG_PYTHON_INITIALIZE_THREADS  PyEval_InitThreads()
>       |                                                                 ^
> controller310.cpp:27431:3: note: in expansion of macro ‘SWIG_PYTHON_INITIALIZE_THREADS’
> 27431 |   SWIG_PYTHON_INITIALIZE_THREADS;
>       |   ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
> In file included from /usr/include/python3.10/Python.h:130,
>                  from controller310.cpp:182:
> /usr/include/python3.10/ceval.h:122:37: note: declared here
>   122 | Py_DEPRECATED(3.9) PyAPI_FUNC(void) PyEval_InitThreads(void);
> ```
> 
> Is it being addressed in this PR already?

Yes. It is fixed in this PR.
There are a couple of other fixes related to Python 3.10 that I will also port to the master branch.

ad-daniel commented on scripts/install/linux_compilation_dependencies.sh:
```diff
@@ -11,6 +11,8 @@ apt install --yes git lsb-release cmake swig libglu1-mesa-dev libglib2.0-dev lib
 if [[ $UBUNTU_VERSION == "20.04" ]]; then
        apt install --yes libzip5
+elif [[ $UBUNTU_VERSION == "22.04" ]]; then
+       apt install --yes libzip4
```
Is it normal we install libzip4 on 22, and libzip5 on 20?

stefaniapedrazzi replied:
Yes, this is the default library in Ubuntu 22.04.

ad-daniel reviewed (changes requested):
I've tested the installation procedure on a virtualbox and it appears to be working nicely, SUMO included. Mostly very minor details/questions. Thank you

stefaniapedrazzi commented on docs/guide/installation-procedure.md:
```diff
@@ -142,7 +141,7 @@ However, the sand-boxing constraints of snaps yield the following limitations:
 
 The download is significantly bigger as it includes all the dependencies of Webots (ffmpeg, Python, C++ and Java compilers, etc.).
-For Webots R2019b revision 1, the download size of the snap is 1.8GB compared to 1.3GB of the Debian and tarball packages.
+For Webots R2022a, the download size of the snap is 821MB compared to 290MB of the Debian and tarball packages.
```
Yes, but we don't know exactly which size the package will be yet.
And we are not even interested in updating this information for every release.
I think it is better to use exact old data than data related to R2022b that we will have to update again later or will be wrong.

stefaniapedrazzi commented on docs/guide/system-requirements.md:
```diff
@@ -13,11 +13,11 @@ Webots works well on all the graphics cards included in fairly recent Apple comp
 Webots is provided for Linux 64 (x86-64) systems.
-Webots doesn't run on Ubuntu versions earlier than 18.04.
+Webots doesn't run on Ubuntu versions earlier than 20.04.
 - Windows: Webots runs on Windows 10 and Windows 8.1 (64-bit versions only).
 - Mac: Webots runs on macOS 10.15 "Catalina" and 10.14 "Mojave".
```
I would update this in a different PR.

stefaniapedrazzi commented:
I think I addressed all the comments. I would update the documentation of supported macOS and Windows version in a different PR, if it is the case we should also change it in the master/released branch.

ad-daniel reviewed (approved):
Thank you

omichel commented on scripts/packaging/Makefile:
```diff
@@ -106,7 +106,7 @@ install: all
 $(WEBOTS_DISTRO_EXE): webots_distro.c
 	+@echo "# compiling webots_distro.c"
-	+@$(CC) $(CFLAGS) -Wall -o $@ webots_distro.c $(LIBRARIES)
+	+@$(CC) $(CFLAGS) -Wno-deprecated-declarations -Wall -o $@ webots_distro.c $(LIBRARIES)
```
I assume that deprecated declarations silenced here are specific to Ubuntu 22.04. If so, I would put this inside the Ubuntu 20.04 `CFLAGS`.

omichel commented on tests/test_suite.py:
```diff
@@ -30,6 +30,11 @@
 
 from command import Command
 
+is_ubuntu_22_04 = False
+if sys.platform == 'linux':
+    result = subprocess.run(['lsb_release', '-sr'], stdout=subprocess.PIPE)
+    is_ubuntu_22_04 = result.stdout.decode().strip() == '22.04'
+
```
This is just a minor style issue: there is a double initialization here, instead we should set `is_ubuntu_22_04` to `False` in a `else` statement.
